[scroll-animations] make `animation-timeline`, `animation-range-start` and `animation-range-end` reset-only longhands of the `animation` shorthand

graouts opened the pull request:
#### 91fb05af09f927320fd7cf97e17f575056b53443
<pre>
[scroll-animations] make `animation-timeline`, `animation-range-start` and `animation-range-end` reset-only longhands of the `animation` shorthand
<a href="https://bugs.webkit.org/show_bug.cgi?id=287398">https://bugs.webkit.org/show_bug.cgi?id=287398</a>

Reviewed by Tim Nguyen.

The CSS WG resolved in <a href="https://github.com/w3c/csswg-drafts/issues/6946#issuecomment-1233190360">https://github.com/w3c/csswg-drafts/issues/6946#issuecomment-1233190360</a> to make the
new CSS properties introduced to support Scroll-driven Animations – `animation-timeline`, `animation-range-start`
and `animation-range-end` – reset-only longhands of the `animation` shorthand.

To that end, we add those properties to the `&quot;longhands&quot;` property for `animation` in `CSSProperties.json`
and deal with them when relevant:

- in `CSSPropertyParser.cpp` we set `nullptr` values for those properties to reset any value previously set
and ensure those properties have the right initial values for serialization purposes.

- in `ComputedStyleExtractor.cpp` we return an empty value whenever we attempt to serialize the `animation`
shorthand in the computed style when one of those reset-only longhands was otherwise set.

- in `ShorthandSerializer.cpp` we do similar work when serializing the specified value.

* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/parsing/animation-shorthand-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/animation-shorthand-expected.txt:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::animationShorthandValue):
* Source/WebCore/css/ShorthandSerializer.cpp:
(WebCore::ShorthandSerializer::serializeLayered const):
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::initialValueForLonghand):
(WebCore::consumeAnimationValueForShorthand):
(WebCore::CSSPropertyParser::consumeAnimationShorthand):

Canonical link: <a href="https://commits.webkit.org/290168@main">https://commits.webkit.org/290168@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/260cd38874ab17a6ecbc62335ac1a5202d226962

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/89135 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/8659 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/43960 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/94116 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/39896 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/91186 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/9047 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/16848 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/68682 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26357 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/92137 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/6932 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/80902 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49043 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/6681 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/35279 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/39003 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/77044 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/36266 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/95949 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/16315 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/11955 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/77562 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/16571 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/76690 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/76853 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/21258 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/19868 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/9418 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13972 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/16329 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/21640 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/16070 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/19521 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/17851 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/ShorthandSerializer.cpp:
```diff
@@ -703,9 +704,23 @@ String ShorthandSerializer::serializeLayered() const
                 }
             }
 
+            // If we get to "animation-timeline" and yet haven't encoutered an unskipped value,
```
```suggestion
            // If we get to "animation-timeline" and yet haven't encountered an unskipped value,
```

nt1m commented on Source/WebCore/css/ShorthandSerializer.cpp:
```diff
@@ -703,9 +704,23 @@ String ShorthandSerializer::serializeLayered() const
+            // If we get to "animation-timeline" and yet haven't encoutered an unskipped value,
+            // this means that this "animation" shorthand only has initial values for the non
+            // reset-only longhands and so we cannot serialize it. We only do this when we deal
+            // with multiple layer since the single "none" case would be caught otherwise.
```
```suggestion
            // with multiple layers since the single "none" case would been caught otherwise.
```

graouts committed:
[scroll-animations] make `animation-timeline`, `animation-range-start` and `animation-range-end` reset-only longhands of the `animation` shorthand
https://bugs.webkit.org/show_bug.cgi?id=287398

Reviewed by Tim Nguyen.

The CSS WG resolved in https://github.com/w3c/csswg-drafts/issues/6946#issuecomment-1233190360 to make the
new CSS properties introduced to support Scroll-driven Animations – `animation-timeline`, `animation-range-start`
and `animation-range-end` – reset-only longhands of the `animation` shorthand.

To that end, we add those properties to the `"longhands"` property for `animation` in `CSSProperties.json`
and deal with them when relevant:

- in `CSSPropertyParser.cpp` we set `nullptr` values for those properties to reset any value previously set
and ensure those properties have the right initial values for serialization purposes.

- in `ComputedStyleExtractor.cpp` we return an empty value whenever we attempt to serialize the `animation`
shorthand in the computed style when one of those reset-only longhands was otherwise set.

- in `ShorthandSerializer.cpp` we do similar work when serializing the specified value.

* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/parsing/animation-shorthand-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/animation-shorthand-expected.txt:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::animationShorthandValue):
* Source/WebCore/css/ShorthandSerializer.cpp:
(WebCore::ShorthandSerializer::serializeLayered const):
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::initialValueForLonghand):
(WebCore::consumeAnimationValueForShorthand):
(WebCore::CSSPropertyParser::consumeAnimationShorthand):

Canonical link: https://commits.webkit.org/290168@main
